### ISSUE TITLE
RegexOptions.NonBacktracking fixed-length markers never triggered

### ISSUE THREAD

**Issue by stephentoub** [dotnet/runtime#65532]
Running our regex tests with code coverage shows this for SymbolicRegexMatcher.FindMatch:
<img width="398" alt="image" src="https://user-images.githubusercontent.com/2642209/154585534-2ee47eaa-29cd-45a0-8a94-cca6b79b4a5f.png">
This suggests that there's something very wrong with that logic, as we definitely have tests that should trigger it, and it should either be fixed or all the "watchdog" logic deleted.

cc: @olsaarik, @veanes

**Comment by veanes** [comment]
@olsaarik The watchdog logic may have become something that is no longer triggered because of all the other optimizations. 
If this is the case it should be deleted, it may also cause conflict inside ordered-or. We'll discuss it also next week (Tuesday) on our separate weekly meeting with Olli.

**Comment by veanes** [comment]
I did some debugging. It seems there is a change in the converter code that at some point turned this feature off, so watchdogs are not inserted although they should be inserted:
RegexNodeToSymbolicConverter.cs: 
where the concatenation `topLevel:false` in line 240, I think this is a bug
not sure when or why this happened:

```
                case RegexNodeKind.Capture when node.N == -1:
                    int captureNum;
                    if (_caps == null || !_caps.TryGetValue(node.M, out captureNum))
                        captureNum = node.M;
                    return _builder.MkCapture(Convert(node.Child(0), topLevel: false), captureNum);
```
I believe the correct code should be the following where topLevel is the input parameter of the `Convert ` method, unless there is another reason why topLevel was turned off (related to **captures** maybe?)
@olsaarik could also take a look at this.
```
                case RegexNodeKind.Capture when node.N == -1:
                    int captureNum;
                    if (_caps == null || !_caps.TryGetValue(node.M, out captureNum))
                        captureNum = node.M;
                    return _builder.MkCapture(Convert(node.Child(0), topLevel), captureNum);
```
Watchdogs were only appended after top level concatenations, but here this info is turned off, so that later on, any potential watchdog is never appended even when fixed length is clearly detected.

**Comment by stephentoub** [comment]
That changed here:
https://github.com/dotnet/runtime/pull/65129/files#diff-38a53311d56a4f575c58bf43d19179b87fec1087bfa0b2be1af755d684618bc9L234

**Comment by olsaarik** [comment]
@veanes Good find. This was not intended and that `Convert` call should indeed just pass in topLevel.

**Comment by veanes** [comment]
@olsaarik thanks for confirming this. I'll fix it with my next edits and also add an extra unit test I was using that should clearly trigger it, no matter what other optimizations are in place.

**Comment by stephentoub** [comment]
The value is now being just passed along for a capture:
https://github.com/dotnet/runtime/blob/68fb7fc68cc1af800bee1d38af22b5027bf4ab4e/src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/RegexNodeConverter.cs#L141
but it's still not being hit:
<img width="392" alt="image" src="https://user-images.githubusercontent.com/2642209/155914746-6d66dc95-f899-487e-82ad-24053d03e396.png">

**Comment by stephentoub** [comment]
I was hoping this may have been addressed by #68199, but alas, it was not:
<img width="651" alt="image" src="https://user-images.githubusercontent.com/2642209/164146474-93fcc541-0a8d-4b0d-a995-de663679f443.png">

**Comment by olsaarik** [comment]
This was fixed in https://github.com/dotnet/runtime/pull/69839. Fixed length marker use was also further improved in https://github.com/dotnet/runtime/pull/70217.
